azurerm/internal/services/compute/resource_arm_disk_encryption_set.go

```diff
@@ -0,0 +1,387 @@
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["diskEncryptionSets"]
+
+	future, err := client.Delete(ctx, resourceGroup, name)
+	if err != nil {
+		if response.WasNotFound(future.Response()) {
+			return nil
+		}
+		return fmt.Errorf("Error deleting Disk Encryption Set %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("Error waiting for deleting Disk Encryption Set %q (Resource Group %q): %+v", name, resourceGroup, err)
+		}
+	}
+
+	return nil
+}
+
+func expandArmDiskEncryptionSetKeyVaultAndKeyReference(input []interface{}) *compute.KeyVaultAndKeyReference {
+	if len(input) == 0 {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+
+	sourceVaultId := v["source_vault_id"].(string)
+	keyUrl := v["key_url"].(string)
+
+	result := compute.KeyVaultAndKeyReference{
+		KeyURL: utils.String(keyUrl),
+		SourceVault: &compute.SourceVault{
+			ID: utils.String(sourceVaultId),
+		},
+	}
+	return &result
+}
+
+func expandArmDiskEncryptionSetIdentity(input []interface{}) *compute.EncryptionSetIdentity {
+	if len(input) == 0 {
+		return nil
+	}
+	v := input[0].(map[string]interface{})
+
+	t := v["type"].(string)
+	result := compute.EncryptionSetIdentity{
+		Type: compute.DiskEncryptionSetIdentityType(t),
+	}
+
+	return &result
+}
+
+func flattenArmDiskEncryptionSetKeyVaultAndKeyReference(input *compute.KeyVaultAndKeyReference) []interface{} {
+	if input == nil {
+		return make([]interface{}, 0)
+	}
+
+	result := make(map[string]interface{})
+
+	if keyUrl := input.KeyURL; keyUrl != nil {
+		result["key_url"] = *keyUrl
+	}
+	if sourceVault := input.SourceVault; sourceVault != nil {
+		if sourceVaultId := sourceVault.ID; sourceVaultId != nil {
+			result["source_vault_id"] = *sourceVaultId
+		}
+	}
+
+	return []interface{}{result}
+}
+
+func flattenArmDiskEncryptionSetKeyVaultAndKeyReferenceArray(input *[]compute.KeyVaultAndKeyReference) []interface{} {
+	results := make([]interface{}, 0)
+	if input == nil {
+		return results
+	}
+
+	for _, item := range *input {
+		v := make(map[string]interface{})
+
+		if sourceVault := item.SourceVault; sourceVault != nil {
+			if sourceVaultId := sourceVault.ID; sourceVaultId != nil {
+				v["source_vault_id"] = *sourceVaultId
+			}
+		}
+
+		results = append(results, v)
+	}
+
+	return results
+}
+
+func flattenArmDiskEncryptionSetIdentity(input *compute.EncryptionSetIdentity) []interface{} {
+	if input == nil {
+		return make([]interface{}, 0)
+	}
+
+	result := make(map[string]interface{})
+
+	result["type"] = string(input.Type)
+	if principalId := input.PrincipalID; principalId != nil {
+		result["principal_id"] = *principalId
+	}
+	if tenantId := input.TenantID; tenantId != nil {
+		result["tenant_id"] = *tenantId
+	}
+	return []interface{}{result}
```

Comment:
(same here) can we ensure these always have a value as above

Reply:
I have modified this to make sure all these field would have a value, please have a look~